10-Profiling.md

```diff
@@ -112,516 +117,573 @@ System time
+However, `gperftool` can do that for you.
+`gperftool` samples your program (by default, 100 times per second) as it runs to tell you how much time you're spending in function calls.
+We'll use `kcachegrind` to display `gperftool`'s output in a GUI and discuss what its results mean.
 
 First, let's look at an example program.
-In this program, we have a few different functions.
-Each one has a for-loop that just wastes time for the sake of example.
+In this program, we have a few different functions:
+- Two different methods for computing the $n^{th}$ Fibonnaci number (`fib_iterative` and `fib_recursive`)
+- Two functions that repeatedly compute the $20^{th}$ Fibonnaci number (`do_iterative` and `do_recursive`)
+- `main`, which calls `fib_iterative` 200,000 times and `fib_recursive` 200 times
 
-~~~cpp
+~~~{.cpp .numberLines}
 #include <iostream>
 using namespace std;
 
-void new_func1()
+long fib_iterative(const long n)
 {
-    cout << "Inside new_func1" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    return;
+  long fib = 1;
+  long prev_fib = 0;
+
+  for(long i = 0; i < n; i++)
+  {
+    long tmp = fib;
+    fib += prev_fib;
+    prev_fib = tmp;
+  }
+
+  return fib;
 }
 
-void func1()
+void do_iterative(const int count)
 {
-    cout << "Inside func1" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    new_func1();
-    return;
+  volatile long fib;
+  cout << "Iterative: ";
+  for(int i = 0; i < count; i++)
+    fib = fib_iterative(20);
+  cout << fib << endl;
 }
 
-void func2()
+long fib_recursive(const long n)
 {
-    cout << "Inside func2" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    return;
+  if(n <= 1) return 1;
+  return fib_recursive(n-1) + fib_recursive(n-2);
 }
 
-int main(void)
+void do_recursive(const int count)
 {
-    cout << "Inside main" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    func1();
-    func2();
-    return 0;
+  volatile long fib;
+  cout << "Recursive: ";
+  for(int i = 0; i < count; i++)
+    fib = fib_recursive(20);
+  cout << fib << endl;
+}
+
+int main()
+{
+  const int count = 200;
+
+  do_iterative(count * 1000);
+
+  do_recursive(count);
+  return 0;
 }
 ~~~
 
-In order to use `gprof` we have to pass an additional flag to `g++`.
-The `-pg` flag tells `g++` to record profile information whenever our compiled program runs.
-The generated file, called `gmon.out`, contains the information that is interpreted by `gprof`.
+In order to use `gperftool`, we have to link its library with our code using the `-lprofiler` flag.
+We'll also want to use the `-g` flag so that `kcachegrind` can show us our source code when we're viewing our profiling output!
 
 Let's compile the program above.
-We'll assume it's called `main.cpp`.
+We'll assume it's called `fibonnaci.cpp`.
 
-~~~shell
+~~~
 $ ls
-main.cpp
-$ g++ -pg -o main main.cpp
+fibonnaci.cpp
+$ g++ -g fibonnaci.cpp -lprofiler -o fibonnaci
```

Comment:
change this line to `g++ -g -lprofiler -o fibonnaci fibonnaci.cpp` so it matches the format of the other compilations in this chapter.